### DISCUSSION TITLE
CD pipelines & modelling environment promotion

### DISCUSSION THREAD

**Discussion by janakerman** [fluxcd/flux2#1011] [Q&A]
I'm relatively new to Kubernetes + Flux, and I'm struggling to model a deployment pipeline with promotion between environments. It's a pretty simple dev -> test -> production setup with some fun twists:

1. The release cycle of each service should be independent (i.e promote foo to test, promote bar to test, promote bar to production)
2. There are actually 3 clusters per environment that stay in sync but should be released independently (3 test + 3 production clusters)
3. Changes should go to environments in strict order (dev -> test (all test clusters) -> prod (all prod clusters))
4. Releases should be gated (that is, test & production releases should be manually initiated)
5. The end goal is to automate promotion in a CD pipeline. Development manifests can be modified by humans, but _promotion_ of those changes to the following environments shouldn't require humans to manually modify/move manifests.

I've seen a few automation approaches discussed:
1. Using semantic versioning tags with patterns for different environments and let the ImageUpdateController handle updates [here](https://github.com/fluxcd/flux2/discussions/715) & [here](https://toolkit.fluxcd.io/guides/flux-v1-automation-migration/#how-to-use-semver-image-tags)
    * I imagine the intention is that a CD pipeline would re-tag an image for each environment to trigger deployment to that env?
2. Update each environments Kustomize overlay and commit for each environment's stage [here](https://github.com/fluxcd/flux2/discussions/383) & [here](https://github.com/gitopsrun/webapp-frontend/blob/master/.github/workflows/staging.yaml#L34-L49).

I'm not sure if I'm mixing two concepts (image tag updates & more complex manifest updates), but neither of these approaches completely solve the 'promotion' of a set of manifests between environments. Both only addresses the update of images, and with Kustomize comes with the additional downside that any change to a base would immediately affect production environments, before being proved out in other environments.

I've described some options I'm considering below - it's a bit of an essay so thanks in advance if you make it to the end.

---

### Option 1: Use a branch per cluster and merge between branches

This is quite nice as it uses plain git commands to promote changes (i.e merge), but doesn't work due to fun twist 1 (service releases should be independent) as a merge bring all changes of the source branch to the target branch. 

Once you've ruled out using git merge, I don't see much advantage to this approach over the next options. Also, having manifests spread over 7+ branches doesn't feel very scalable from a visibility perspective, but that seems solveable.

### Option 2: Duplicating manifests

<details>
    <summary>I could completely duplicate the manifests for each service in each environment.
</summary>

```
clusters/
  dev/
    foo/
      helm-release.yaml
  test-1/
    foo/
      ...
  test-2/
  test-3/
  prod-1/
  prod-2/
  prod-3/
...
```

</details>

I could then promote a service's manifests to the next environment by replacing the contents in the next environments directory and raising a PR. This supports fun twist #4, stacks should be independently released. The result is a PR that acts as my deployment gate, with a nice diff of exactly what will change. Service updates are committed to `dev/*` manifests by the ImageUpdateController as per option 3 and engineers modifying manifests directly (i.e modify existing manifests, add new services).

I've written a simple tool that automates this 'promotion' as a proof of concept and it works quite well. You can do `promote --service foo --target-env test` and it'll raise the promotion PR copying the manifests over. I've also implemented some basic logic to close superceeded PRs and could also implement any logic/preconditions that might come along in the future (i.e all test stacks must be updated before any production stack).

The lack of DRY here does feel wrong, but **my gut feeling is that this is OK, as if the state of the individual clusters is different, then differences in the manifest folder if each environment cluster is understandable.** 

The drawback of this approach is that it rules out Kustomize as a way of keeping configuration DRY, mainly for the same reasons as described above, but also due to the added complexity around finding referenced bases and promoting them alongside the service's manifests.  My thoughts are that I could manage variances in environments through config maps which don't need any promotion mechanism and are just directly applied to their respective cluster, but I'm not sure if this is realistic.

```
...
config/
  dev/
  test-1/
  test-2/
  test-3/
...
```

This could fit quite nicely into a CD pipeline per service. One example:
* Foo service pipeline monitors manifest repo for changes in `dev/foo`.
* Forks and raises PR to promote to `test-1/foo`, `test-2/foo`, and `test-3/foo`, pipeline waits for PRs to merge.
* Joins and raises PR to promote to `prod-1/foo`, `prod-2/foo` and `prod-3/foo`, pipeline waits for PRs to merge.

### Option 3:  Single manifest source of truth pushes to cluster repo

<details>
    <summary>Have a single set of manifests in one repository, and a separate repository that represents the cluster state.</summary>

`manifest-repo`
```
services/
  foo/
    helm-release.yaml
config/
  dev/
  test-1/
  test-2/
```

`cluster-repo`:
```
  dev/
    foo/
      helm-release.yaml
    config/
  test-1/
  test-2/
  test-3/
  prod-1/
  prod-2/
  prod-3/
```

</details>

As per option 2, service updates are committed to the `manifest-repo` by the ImageUpdateController and engineers modifying manifests directly. A change in the manifest repository triggers a pipeline for that commit:
* commits the manifests and cluster specific config to `dev/` folder of the `cluster-repo` then;
* opens PRs to the test clusters and waits for merge then;
* opens PRs to the prod clusters and waits for merge

---

Both option 2 & 3 both seem workable, with the main difference being where the progress of promotion for a given service is stored (with option 2 using the previous environments folder, and option 3 using a stateful CD pipeline).

I feel I may be adding unnecessary complexity by trying to manage manifests in a single repo (which ruled out option 1). I've also considered using a [multi-tenancy approach](https://github.com/fluxcd/flux2-multi-tenancy) with a repo per service, but my main concern is visibility, with 60+ services. Splitting some other way (e.g a repo per team) doesn't really solve the problem of individual service promotion.

I have in the back of my mind that I may be trying to fit a square peg into a round hole, but I still think following a Gitops methodology with Flux provides some great benefits. 

Am I overcomplicating this? Has anyone done something similar or had a similar use case? Care to share your experiences/any pitfalls you encountered?

Thank you! 🙏

**Comment by saiyen2002** [comment]
I to am having the same challenges. being new to gitops and flux, and coming from background of promoting dev code into master, I do not see how FluxCD fits this model apart from having to duplicate manifests, which I think is error prone

**Reply by kingdonb** [reply]
Between the https://github.com/fluxcd/flux2-kustomize-helm-example and https://github.com/fluxcd/flux2-multi-tenancy there are a few good examples here of using Kustomize with bases:

https://github.com/fluxcd/flux2-multi-tenancy/blob/main/tenants/production/kustomization.yaml#L4

This is an example of how you can provide one common base for two environments and capture the difference between them in a patch, applied to one or the other environment.

(Compared to how you're describing promotion from dev-> production, I fully realize this is upside down and inside out from how development is actually wanted to be done...)

The best approach which I have understood so far is to (1) provide the base manifests in an app-specific repository, where they are (2) versioned with semantic version git tags, and then (3) pull the release version into the cluster config repository, pinned to a specific tag.

You can point your dev environment toward the main branch of the app repo, and re-point it at any feature branches that you wish to test along the way... patch the latest dev image tag version into those manifests, (probably in the config repo to avoid annoying noise in the app repo which always commits after you)

Point your staging environment towards a tag such as 1.0.0-rc1, and point your production environment towards a tag such as 1.0.1 when it's ready. Automating this part is tricky, as there are several automations to choose from. You can use a `GitRepository` with a `semver` in the spec, like this one: https://github.com/kingdonb/github-actions-demo/blob/89fc2f02880dbe5fce1caec63f53da60f3438ae7/apps/podinfo/ex4-atomic-git-semver-tag-automation/gitrepo.yaml#L10

That might be the most effective way to do this, but it lacks a critical requirement that many users will want: I need to see a commit in the configuration repository whenever the version in production or staging has been changed.

This requirement will be much easier to satisfy when OCI configuration sources are supported in Flux. I'm afraid it's difficult to make it clear exactly why this helps so much, and what the alternative is now absent this support, but I'll give it a shot...

(One way I can recommend to satisfy this requirement uses a haphazard assemblage of ConfigMap and ImageUpdateAutomation, with an obvious kludge that infers "the GitRepository tag must exist, because an Image of that tag version exists, (and image tags are only generated when the CI observes the gitrepository tag) – this kludge will no longer be necessary after OCI lands, but for now it's one of two options, and perhaps one option is better or worse for you depending on your specific requirements here.

A better option is close on the horizon. Please stay tuned for Flux OCI news in the coming months, or please feel free to badger me in CNCF\#flux for more specific, direct and further elaborated examples of the assemblage I've described here.
